[GSB] NFC: Improve performance with local DependentMemberType caching

davezarzycki opened the pull request:
This hack is a 3.5% improvement for Swift.swiftmodule (release, no asserts). Your milage may vary.

@DougGregor – It seems like the GSB could try harder to avoid banging on the global DependentMemberTypes DenseMap. What do you think?

CC: @slavapestov @jrose-apple

slavapestov commented:
I think bigger gains can be had by algorithmic optimizations further upstream so that the GSB can do less work overall.

jrose-apple commented:
I don't disagree but I also don't want to rely on getting those. This sounds like it's still worth it for now, even if it's a bit oblique.

davezarzycki commented:
Of course. In the mean time, what do people think of optimization hacks like this?

davezarzycki committed:
[GSB] NFC: Precompute or cache DependentMemberType when possible

Calling DependentMemberType::get() repeatedly pollutes the processor
caches because the global hash table can be quite large. With this
change, we either precompute DependentMemberTypes or cache them on
demand. This change makes the release/no-assert build of
Swift.swiftmodule 4.1% faster.

davezarzycki commented:
The updated commit is now 4.1% faster.

@swift-ci please test

davezarzycki commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -2935,7 +2942,7 @@ Type GenericSignatureBuilder::PotentialArchetype::getDependentType(
-      return DependentMemberType::get(parentType, assocType);
+      return getResolvedDependentMemberType(parentType);
 
     return DependentMemberType::get(parentType, getNestedName());
```
Oh, I forgot to include this, but this line is dead code. Why? Because potential archetypes with parents always have an associated type, therefore `getResolvedType()` always returns a non-null pointer.

DougGregor replied:
Ahh, this is a vestige from a time that was not true, when we still had potential archetypes that were as-yet unresolved to an associated type.

davezarzycki commented:
@swift-ci please test linux

davezarzycki commented:
@swift-ci please test compiler performance

DougGregor reviewed (approved):
I don't see the algorithm wins any time in the near future, and 4% is definitely a worthwhile improvement.

gottesmm commented:
@davezarzycki this is causing massive leaks on the Linux bots. The LSAN bot is seeing just one leaks spot leaking almost 1 gig of memory. See:

https://ci.swift.org/job/oss-swift-incremental-RA-lsan-linux-ubuntu-18_04/7285/console

We should revert.

davezarzycki commented:
Crud. Do you have magic revert powers? Or do you also need to wait for CI to "test" the revert? I can revert this if you want.

davezarzycki commented:
And are the leaks genuine or just "abandoned" (to use Apple's terminology for effectively but not literally unreachable memory)?

davezarzycki commented:
@DougGregor – Any idea why the destructor might not be called?

davezarzycki commented:
FYI - #23109

gottesmm commented:
@davezarzycki I don't think it is abandoned. It says it is a direct leak. That says that it isn't just about destructors that is called an indirect leak in LSAN speak [I am assuming by destructor you were referring to storing a SmallVector or something in a bump ptr allocated struct and not calling the destructor.

jrose-apple commented:
Ah, I guess we all forgot GSBs are persisted. We could clear the maps on finalization, maybe?

davezarzycki commented:
Ya, I looked into that but I got distracted. Unless DenseMap promises to "free()" its backing store during `clear()`, it would need to be destroyed, not just cleared during GSB finalization (like the potential archetypes).

jrose-apple commented:
It looks like there's a `shrink_and_clear`, or you could do the `swap`-and-destroy trick, or you could make it Optional.
